src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java

```diff
@@ -700,7 +700,7 @@ void shutdown(Throwable t) {
         Throwable initialCause = this.cause;
         if (initialCause == null) this.cause = t;
         client2.deleteConnection(this);
-        List<Stream<?>> c = new LinkedList<>(streams.values());
+        List<Stream<?>> c = new ArrayList<>(streams.values());
```

Comment:
Why can't we dispense with this copy completely, and instead just iterate `streams.values()`?

Reply:
Good point Aleksey. I guess that the original intent was to avoid `ConcurrentModificationException`. However I see that `streams` is a `ConcurrentHashMap` now. So maybe we could dispense of the copy:

`    private final Map<Integer,Stream<?>> streams = new ConcurrentHashMap<>();`

Could be worth changing the type from `Map<Integer,Stream<?>>` to `ConcurrentMap<Integer,Stream<?>>` in the declaration above if we decide to get rid of the copy though.

Reply:
Taking a look at this presently

Reply:
Although Alexey's proposal does not change the behavior, it raises a question (to be explored separately): can a stream be added to that map while `Http2Connection.shutdown` is in progress? If this is the case, then the method is racy.

Reply:
Even if that happened , it should not matter as the connection will be (or is) closed anyway. I don't believe that's the kind of scenario that the copy was trying to prevent. Closing a stream will definitely remove it from the map though, so my guess that is where the CME would happen if no precaution was taken.